### PR TITLE
feat #41: truncate object keys

### PR DESCRIPTION
**Pull request by alemagio**

### CODE REVIEW

**Comment by Qard** [comment]
I've discussed with the team and we decided on _not_ doing the index suffix thing as that could result in non-determinism in the data over time. It could potentially cause a mapping explosion in Elasticsearch, which we don't want. Instead, we'll just let the conflict happen and let the keys overwrite each other. If the user wants to avoid data loss it'll be on them to ensure their custom data doesn't use keys that will conflict.

File: lib/truncate.js [hunk 1]
```diff
@@ -284,3 +300,29 @@ function onObject (value, key, path, isNew) {
...
+    return value
+  }
+  let result = value
+  const keys = Object.keys(result)
+  const truncatedKeys = keys.map(k => keywords.includes(k) ? k : truncate(k, max))
+  const noDuplicates = incrementalDuplicates(truncatedKeys)
+
+  keys.forEach((k, index) => {
+      const {[k]: kValue, ...rest} = result
+      result = {[noDuplicates[index]]: truncateCustomKeys(kValue, max, keywords), ...rest}
```
[the remark this answers is not in the document]
Comment (alemagio) [reply lib/truncate.js]: I thought that spread operator was faster than forcing garbage collector with "delete".
Good to know it is not :sweat_smile:

**Comment by alemagio** [comment]
I also removed the part of the test related to the key collision.

**Comment by alemagio** [comment]
> I've discussed with the team and we decided on _not_ doing the index suffix thing as that could result in non-determinism in the data over time. It could potentially cause a mapping explosion in Elasticsearch, which we don't want. Instead, we'll just let the conflict happen and let the keys overwrite each other. If the user wants to avoid data loss it'll be on them to ensure their custom data doesn't use keys that will conflict.

I removed the test related to the keys collision

File: lib/truncate.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Qard) [reply lib/truncate.js]: If you're concerned about GC, you could do `result[k] = undefined` instead.

[the remark this answers is not in the document]
Comment (alemagio) [reply lib/truncate.js]: No, it's fine, i just take the chance to study a bit more how spread operator works behind the scene.
Thanks

**Comment by Qard** [comment]
jenkins run the tests please

**Comment by Qard** [comment]
This seems to be failing the lint step in the CI. Can you fix the issues listed here: https://travis-ci.org/elastic/apm-nodejs-http-client/jobs/601375026#L218-L229

Also, you will need to fix the commit title format to pass the commit message lint. Can you squash the commits into a single commit titled "feat(truncate): limit object key length" please?

**Comment by Qard** [comment]
Landed and released in elastic-apm-http-client 9.1.0.

**Pull request merged**